## Module 2 HW

In case you don't get one option exactly, select the closest one

For the homework, we'll be working with the green taxi dataset located here:

https://github.com/DataTalksClub/nyc-tlc-data/releases/tag/green/download

To get a wget-able link, use this prefix (note that the link itself gives 404):

https://github.com/DataTalksClub/nyc-tlc-data/releases/download/green/

Assignment
So far in the course, we processed data for the year 2019 and 2020. Your task is to extend the existing flows to include data for the year 2021.

---

As a hint, Kestra makes that process really easy:

You can leverage the backfill functionality in the scheduled flow to backfill the data for the year 2021. Just make sure to select the time period for which data exists i.e. from 2021-01-01 to 2021-07-31. Also, make sure to do the same for both yellow and green taxi data (select the right service in the taxi input).
Alternatively, run the flow manually for each of the seven months of 2021 for both yellow and green taxi data. Challenge for you: find out how to loop over the combination of Year-Month and taxi-type using ForEach task which triggers the flow for each combination using a Subflow task.

---

### Question 1

Within the execution for Yellow Taxi data for the year 2020 and month 12: what is the uncompressed file size (i.e. the output file yellow_tripdata_2020-12.csv of the extract task)?

Answer: `134.5 MiB`

---

### Question 2

What is the rendered value of the variable file when the inputs taxi is set to green, year is set to 2020, and month is set to 04 during execution?

First, I edited the inputs when executing the flow...

<img src="supplements/images/execute.png"
     alt="Here is where I chose the new inputs."
     width="550">

Then, I went to the Overview panel to view the name of the file...

<img src="supplements/images/file_name.png"
     alt="Here is where I found the file name."
     width="550">

Answer: `green_tripdata_2020-04.csv`

---

### Question 3

How many rows are there for the Yellow Taxi data for all CSV files in the year 2020?

I manually ran Flow #4 for each month in 2020, then queried the database as shown below:

In [30]:
import pandas as pd
from sqlalchemy import create_engine, text

In [31]:
engine = create_engine('postgresql://root:root@localhost:5432/ny_taxi')

In [32]:
query = text("""
SELECT COUNT(*) FROM yellow_tripdata WHERE filename LIKE 'yellow_tripdata_2020%';
""")

In [33]:
with engine.connect() as conn:
    df_result = pd.read_sql(query, conn)
    print(df_result)

      count
0  24648499


Answer: `24,648,499`

---

### Question 4

How many rows are there for the Green Taxi data for all CSV files in the year 2020?

I manually ran Flow #4 for each month in 2020, then queried the database as shown below:

In [34]:
query = text("""
SELECT COUNT(*) FROM green_tripdata WHERE filename LIKE 'green_tripdata_2020%';
""")

In [35]:
with engine.connect() as conn:
    df_result = pd.read_sql(query, conn)
    print(df_result)

     count
0  1734051


Answer: `1,734,051`

---

### Question 5

How many rows are there for the Yellow Taxi data for the March 2021 CSV file?

First, I simply edited the YAML for Flow #4 to reference 2021 instead of 2019 & 2020:

<img src="supplements/images/yaml_flow.png"
     alt="Here is where I edited the flow."
     width="550">

Then I queried the database as before:

In [38]:
query = text("""
SELECT COUNT(*) FROM yellow_tripdata WHERE filename LIKE 'yellow_tripdata_2021-03%';
""")

In [39]:
with engine.connect() as conn:
    df_result = pd.read_sql(query, conn)
    print(df_result)

     count
0  1925152


Answer: `1,925,152`

---

### Question 6

How would you configure the timezone to New York in a Schedule trigger?

This is clearly delineated in Kestra's docs: https://kestra.io/docs/workflow-components/triggers/schedule-trigger

<img src="supplements/images/timezone.png"
     alt="Here is how timezones are defined in Kestra."
     width="550">

Answer: `Add a timezone property set to America/New_York in the Schedule trigger configuration.`

---